R/netzschleuder.R

```diff
@@ -0,0 +1,205 @@
+          "i" = "see {.url {collection_url}}"
+        )
+      )
+    }
+    raw[["analyses"]] <- raw[["analyses"]][[net_ident[2]]]
+    raw[["nets"]] <- raw[["nets"]][idx]
+    raw
+  }
+}
+
+#' @rdname netzschleuder
+#' @export
+ns_df <- function(name, token = NULL) {
+  if (!requireNamespace("minty", quietly = TRUE)) {
+    cli::cli_abort(
+      "The package `minty` is needed for this function. Please install it."
+    )
+  }
+  meta <- ns_metadata(name)
+  net_ident <- resolve_name(name)
+
+  zip_url <- sprintf(
+    "net/%s/files/%s.csv.zip",
+    net_ident[1],
+    net_ident[2]
+  )
+
+  temp <- tempfile(fileext = "zip")
+  download_file(zip_url, token = token, file = temp)
+
+  zip_contents <- utils::unzip(temp, list = TRUE)
+
+  edge_file_name <- grep("edge", zip_contents$Name, value = TRUE)
+  node_file_name <- grep("node", zip_contents$Name, value = TRUE)
+  gprops_file_name <- grep("gprops", zip_contents$Name, value = TRUE)
+
+  edges_df_raw <- utils::read.csv(unz(temp, edge_file_name))
+  edges_df <- suppressWarnings(minty::type_convert(edges_df_raw))
+  source_loc <- grep("source", names(edges_df))
+  target_loc <- grep("target", names(edges_df))
+
+  # netzschleuder uses 0-indexing, igraph uses 1-indexing
+  edges_df[[source_loc]] <- edges_df[[source_loc]] + 1
+  edges_df[[target_loc]] <- edges_df[[target_loc]] + 1
+  names(edges_df)[c(source_loc, target_loc)] <- c("from", "to")
+
+  nodes_df_raw <- utils::read.csv(unz(temp, node_file_name))
+  #suppress warning if no character columns found
+  nodes_df <- suppressWarnings(minty::type_convert(nodes_df_raw))
+  names(nodes_df)[1] <- "id"
+
+  # netzschleuder uses 0-indexing, igraph uses 1-indexing
+  nodes_df[["id"]] <- nodes_df[["id"]] + 1
+  if ("X_pos" %in% names(nodes_df)) {
+    pos_array <- gsub("array\\(\\[|\\]|\\)", "", nodes_df[["X_pos"]])
+    split_coords <- strsplit(pos_array, ",")
+
+    x_vals <- vapply(
+      split_coords,
+      function(x) as.numeric(trimws(x[1])),
+      numeric(1)
+    )
+    y_vals <- vapply(
+      split_coords,
+      function(x) as.numeric(trimws(x[2])),
+      numeric(1)
+    )
+
+    nodes_df[["X_pos"]] <- NULL
+    nodes_df[["x"]] <- x_vals
+    nodes_df[["y"]] <- y_vals
+  }
+
+  #gprops file is not always a valid csv file. In that case, simply read it as text
+  gprops_df <- readLines(unz(temp, gprops_file_name))
+
+  on.exit(unlink(temp))
+
+  list(nodes = nodes_df, edges = edges_df, gprops = gprops_df, meta = meta)
+}
+
+#' @rdname netzschleuder
+#' @export
+ns_graph <- function(name, token = NULL) {
+  graph_data <- ns_df(name, token = token)
+  directed <- graph_data$meta[["analyses"]][["is_directed"]]
+  bipartite <- graph_data$meta[["analyses"]][["is_bipartite"]]
+
+  g <- igraph::graph_from_data_frame(
+    graph_data$edges,
+    directed = directed,
+    vertices = graph_data$nodes
+  )
+
+  if (bipartite) {
+    types <- rep(FALSE, igraph::vcount(g))
+    types[graph_data$nodes$id %in% graph_data$edges[, 1]] <- TRUE
```

Comment:
Are we looking at the first column in a data frame? Can be shortened, I think:

```suggestion
    type <- (graph_data$nodes$id %in% graph_data$edges[[1]])
```

Should this be `[[2]]` instead? My reasoning here is that the first vertex mentioned in an edge often is the source vertex, and with bipartite directed graphs it might also make sense to have the "first" vertex set correspond to the source vertices. Now I don't know if `type == TRUE` or `FALSE` is the "first" vertex set.